[inductor] Add single pass "var_unnormalized" reduction_type

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #102737
* __->__ #102486

This is a bit inefficient because it computes the mean and throws it
away since ir.Reduction nodes only have 1 output. However, the mean
can at least be scheduled into the same loop as the variance now since
there is no data dependency. Thus we can take fewer passes over the
data.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @ngimel @yf225 @chenyang78 @kadeng

peterbell10 committed:
[inductor] Add single pass "var" reduction_type

This is a bit inefficient because it computes the mean but thows it
away since ir.Reduction nodes only have 1 output. However, the mean
can at least be scheduled into the same loop as the variance now since
there is no data dependency. Thus we can take fewer passes over the
data.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Add single pass "var" reduction_type"


This is a bit inefficient because it computes the mean but thows it
away since ir.Reduction nodes only have 1 output. However, the mean
can at least be scheduled into the same loop as the variance now since
there is no data dependency. Thus we can take fewer passes over the
data.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Add single pass "var" reduction_type"


This is a bit inefficient because it computes the mean but thows it
away since ir.Reduction nodes only have 1 output. However, the mean
can at least be scheduled into the same loop as the variance now since
there is no data dependency. Thus we can take fewer passes over the
data.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Add single pass "var" reduction_type"


This is a bit inefficient because it computes the mean but thows it
away since ir.Reduction nodes only have 1 output. However, the mean
can at least be scheduled into the same loop as the variance now since
there is no data dependency. Thus we can take fewer passes over the
data.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

ngimel commented:
Should we modify ir.Reduction to have multiple outputs?
Or could we remedy this in at the codegen level, by using `tl.reduce` that can now do multiple reductions at once, thanks to you?

peterbell10 commented:
> Should we modify `ir.Reduction` to have multiple outputs?

I tried going in this direction at first, but it's kind of baked into the IR format that there is one output per loop. Changing that seemed to be quite a large effort. Thankfully, my benchmarks show that this is still a huge win even if we do recompute the mean because the memory bandwidth is used so much more efficiently.

> Or could we remedy this in at the codegen level, by using `tl.reduce` that can now do multiple reductions at once, thanks to you?

Yeah merging multiple reduction ops into a single `tl.reduce` call could be interesting to explore. I do think it would be nicer if triton could do this automatically though. I think there are trade-offs in terms of instruction scheduling and shared memory usage that are a lot harder to reason about from inductor's side.

peterbell10 committed:
Update on "[inductor] Add single pass "var" reduction_type"


This is a bit inefficient because it computes the mean but thows it
away since ir.Reduction nodes only have 1 output. However, the mean
can at least be scheduled into the same loop as the variance now since
there is no data dependency. Thus we can take fewer passes over the
data.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

Chillee commented on torch/_inductor/lowering.py:
```diff
@@ -3636,15 +3636,48 @@ def var_mean_(x, axis, correction, keepdim, return_mean):
 def var_mean(x, axis=None, *, correction=None, keepdim=False):
-    return var_mean_(
-        x, axis=axis, correction=correction, keepdim=keepdim, return_mean=True
+    if torch.version.hip is not None:
```
torch.version.hip?

ngimel commented:
I'm slightly concerned about the impact this might have on `var_mean` reduction across a small dimension that actually allows persistent reduction (like layer_norm with size 768-1024 or maybe even 2048, though we don't allow persistent reduction for 2048 yet) - in this case persistent "2 pass" reduction might be better because data is loaded only once, and (I think) it requires fewer registers. Can we have different lowerings based on the reduction dimension? Even with dynamic shapes ln reduction dimention should be known at compile time.

peterbell10 committed:
Update on "[inductor] Add single pass "var_unnormalized" reduction_type"


This is a bit inefficient because it computes the mean and throws it
away since ir.Reduction nodes only have 1 output. However, the mean
can at least be scheduled into the same loop as the variance now since
there is no data dependency. Thus we can take fewer passes over the
data.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Add single pass "var_unnormalized" reduction_type"


This is a bit inefficient because it computes the mean and throws it
away since ir.Reduction nodes only have 1 output. However, the mean
can at least be scheduled into the same loop as the variance now since
there is no data dependency. Thus we can take fewer passes over the
data.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Add single pass "var_unnormalized" reduction_type"


This is a bit inefficient because it computes the mean and throws it
away since ir.Reduction nodes only have 1 output. However, the mean
can at least be scheduled into the same loop as the variance now since
there is no data dependency. Thus we can take fewer passes over the
data.

[ghstack-poisoned]

peterbell10 committed:
Limit threads used for triton build on "[inductor] Add single pass "var_unnormalized" reduction_type"


This is a bit inefficient because it computes the mean and throws it
away since ir.Reduction nodes only have 1 output. However, the mean
can at least be scheduled into the same loop as the variance now since
there is no data dependency. Thus we can take fewer passes over the
data.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Add single pass "var_unnormalized" reduction_type"


This is a bit inefficient because it computes the mean and throws it
away since ir.Reduction nodes only have 1 output. However, the mean
can at least be scheduled into the same loop as the variance now since
there is no data dependency. Thus we can take fewer passes over the
data.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Add single pass "var_unnormalized" reduction_type"


This is a bit inefficient because it computes the mean and throws it
away since ir.Reduction nodes only have 1 output. However, the mean
can at least be scheduled into the same loop as the variance now since
there is no data dependency. Thus we can take fewer passes over the
data.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Add single pass "var_unnormalized" reduction_type"


This is a bit inefficient because it computes the mean and throws it
away since ir.Reduction nodes only have 1 output. However, the mean
can at least be scheduled into the same loop as the variance now since
there is no data dependency. Thus we can take fewer passes over the
data.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Add single pass "var_unnormalized" reduction_type"


This is a bit inefficient because it computes the mean and throws it
away since ir.Reduction nodes only have 1 output. However, the mean
can at least be scheduled into the same loop as the variance now since
there is no data dependency. Thus we can take fewer passes over the
data.

[ghstack-poisoned]

jgong5 reviewed (commented):
Do you have performance benchmarking comparing the the numbers before and after the changes?

peterbell10 committed:
Update on "[inductor] Add single pass "var_unnormalized" reduction_type"


This is a bit inefficient because it computes the mean and throws it
away since ir.Reduction nodes only have 1 output. However, the mean
can at least be scheduled into the same loop as the variance now since
there is no data dependency. Thus we can take fewer passes over the
data.

[ghstack-poisoned]

jgong5 commented on torch/_inductor/codegen/cpp_prefix.h:
```diff
@@ -12,16 +12,116 @@
+
+#if INDUCTOR_USE_VECTOR_TYPES()
+template <typename scalar_t>
+inline at::vec::Vectorized<scalar_t> vec_shuffle_down(at::vec::Vectorized<scalar_t> x, size_t n) {
```
better with permute instruction?

peterbell10 committed:
Update on "[inductor] Add single pass "var_unnormalized" reduction_type"


This is a bit inefficient because it computes the mean and throws it
away since ir.Reduction nodes only have 1 output. However, the mean
can at least be scheduled into the same loop as the variance now since
there is no data dependency. Thus we can take fewer passes over the
data.

[ghstack-poisoned]

peterbell10 committed:
Rebase on "[inductor] Add single pass "var_unnormalized" reduction_type"


This is a bit inefficient because it computes the mean and throws it
away since ir.Reduction nodes only have 1 output. However, the mean
can at least be scheduled into the same loop as the variance now since
there is no data dependency. Thus we can take fewer passes over the
data.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Add single pass "var_unnormalized" reduction_type"


This is a bit inefficient because it computes the mean and throws it
away since ir.Reduction nodes only have 1 output. However, the mean
can at least be scheduled into the same loop as the variance now since
there is no data dependency. Thus we can take fewer passes over the
data.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Add single pass "var_unnormalized" reduction_type"


This is a bit inefficient because it computes the mean and throws it
away since ir.Reduction nodes only have 1 output. However, the mean
can at least be scheduled into the same loop as the variance now since
there is no data dependency. Thus we can take fewer passes over the
data.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Add single pass "var_unnormalized" reduction_type"


This is a bit inefficient because it computes the mean and throws it
away since ir.Reduction nodes only have 1 output. However, the mean
can at least be scheduled into the same loop as the variance now since
there is no data dependency. Thus we can take fewer passes over the
data.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Add single pass "var_unnormalized" reduction_type"


This is a bit inefficient because it computes the mean and throws it
away since ir.Reduction nodes only have 1 output. However, the mean
can at least be scheduled into the same loop as the variance now since
there is no data dependency. Thus we can take fewer passes over the
data.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Add single pass "var_unnormalized" reduction_type"


This is a bit inefficient because it computes the mean and throws it
away since ir.Reduction nodes only have 1 output. However, the mean
can at least be scheduled into the same loop as the variance now since
there is no data dependency. Thus we can take fewer passes over the
data.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Add single pass "var_unnormalized" reduction_type"


This is a bit inefficient because it computes the mean and throws it
away since ir.Reduction nodes only have 1 output. However, the mean
can at least be scheduled into the same loop as the variance now since
there is no data dependency. Thus we can take fewer passes over the
data.

[ghstack-poisoned]

peterbell10 commented:
Looking at the benchmark results, this is a significant win for some specific timm models with dynamic shapes. 

model | change in speedup
----------|---------------------------
adv_inception_v3    | 1.1318 → 1.2497
gluon_inception_v3 | 1.1366 → 1.2473
inception_v3             | 1.1333 → 1.2473
timm_efficientdet | 1.1249 → 1.2690

Presumably the static shape variants are using persistent reductions, so don't benefit from this change.

https://hud.pytorch.org/benchmark/timm_models/inductor_dynamic?startTime=Wed%2C%2007%20Jun%202023%2021%3A10%3A11%20GMT&stopTime=Wed%2C%2014%20Jun%202023%2021%3A10%3A11%20GMT&granularity=hour&mode=training&dtype=amp&lBranch=gh/peterbell10/562/head&lCommit=dda51adf26e627762b00049eb67b80b79eed1f70&rBranch=main&rCommit=593642d1d8d0edec05eaeeca38bb2b8be9b73639

Chillee commented:
IIRC, the other setting where this seemed like it would be a significant win was for channels-last batch norm. That might also be worth spot checking?

peterbell10 committed:
Update on "[inductor] Add single pass "var_unnormalized" reduction_type"


This is a bit inefficient because it computes the mean and throws it
away since ir.Reduction nodes only have 1 output. However, the mean
can at least be scheduled into the same loop as the variance now since
there is no data dependency. Thus we can take fewer passes over the
data.

fixup! [inductor] Update triton pin

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Add single pass "var_unnormalized" reduction_type"


This is a bit inefficient because it computes the mean and throws it
away since ir.Reduction nodes only have 1 output. However, the mean
can at least be scheduled into the same loop as the variance now since
there is no data dependency. Thus we can take fewer passes over the
data.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Add single pass "var_unnormalized" reduction_type"


This is a bit inefficient because it computes the mean and throws it
away since ir.Reduction nodes only have 1 output. However, the mean
can at least be scheduled into the same loop as the variance now since
there is no data dependency. Thus we can take fewer passes over the
data.

[ghstack-poisoned]

lezcano commented on torch/_inductor/codegen/cpp.py:
```diff
@@ -49,6 +49,24 @@
 
 schedule_log = torch._logging.getArtifactLogger(__name__, "schedule")
 
+DTYPE_TO_COMPUTATION_DTYPE = {
+    torch.bfloat16: torch.float,
+    torch.float16: torch.float,
+    **{
+        dtype: dtype
+        for dtype in [
+            torch.bool,
+            torch.float32,
+            torch.float64,
+            torch.int8,
+            torch.int16,
+            torch.int32,
+            torch.int64,
+            torch.uint8,
+        ]
+    },
+}
+
```
This is `_prims_common.get_computation_dtype`.

lezcano commented on torch/_inductor/codegen/triton.py:
```diff
@@ -1380,6 +1402,36 @@ def final_argreduce(buffer, result_var, value, index):
                 """
                 )
                 final_argreduce(self.suffix, result_var, accumulator, accumulator_index)
+            elif reduction_type == "var_unnormalized":
```
It'd be so nice if we could generate these from the `combine_fn`...

lezcano commented on torch/_inductor/lowering.py:
```diff
@@ -3768,56 +3768,66 @@ def _validate_reduction_axis(x, axis):
-            ):
-                new_index[idx] = var
-            return inner_loader(new_index)
+def _make_reduction_inner(x, *, axis, keepdims, dtype, override_return_dtype):
```
why are the diffs so bad :'(

lezcano replied:
Is this part all just code shuffling?

peterbell10 replied:
The diff is much clearer if you ignore whitespace changes,
https://github.com/pytorch/pytorch/pull/102486/files?diff=unified&w=1

lezcano replied:
TIL!

lezcano reviewed (approved):
Cool! I have not reviewed the C++ generation code though. @jgong5 could you have a look at that part?

lezcano commented on torch/_inductor/ir.py:
```diff
@@ -921,56 +968,35 @@ def fn(index):
-                "argmin",
-            }
-            and config.split_reductions
+        # triton doesn't support reduce to single element well, so break it up
```
nit. This comment is repeated a few lines below, and now it's sort of stale, better remove it.

peterbell10 committed:
Update on "[inductor] Add single pass "var_unnormalized" reduction_type"


This is a bit inefficient because it computes the mean and throws it
away since ir.Reduction nodes only have 1 output. However, the mean
can at least be scheduled into the same loop as the variance now since
there is no data dependency. Thus we can take fewer passes over the
data.

[ghstack-poisoned]

peterbell10 committed:
Update on "[inductor] Add single pass "var_unnormalized" reduction_type"


This is a bit inefficient because it computes the mean and throws it
away since ir.Reduction nodes only have 1 output. However, the mean
can at least be scheduled into the same loop as the variance now since
there is no data dependency. Thus we can take fewer passes over the
data.

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge

clee2000 commented:
@pytorchbot revert -m "sorry but this seems to have broken inductor/test_torchinductor.py::CpuTests::test_std_cpu on mac x86 64 machines https://hud.pytorch.org/pytorch/pytorch/commit/7e098f95593240d45d28f040ff53f268ad3d9a93 https://github.com/pytorch/pytorch/actions/runs/5479008241/jobs/9981443710" -c nosignal

clee2000 commented:
@pytorchbot revert -m "sorry but this seems to have broken inductor/test_torchinductor.py::CpuTests::test_std_cpu on mac x86 64 machines https://hud.pytorch.org/pytorch/pytorch/commit/7e098f95593240d45d28f040ff53f268ad3d9a93 https://github.com/pytorch/pytorch/actions/runs/5479008241/jobs/9981443710" -c nosignal

peterbell10 committed:
Update on "[inductor] Add single pass "var_unnormalized" reduction_type"


This is a bit inefficient because it computes the mean and throws it
away since ir.Reduction nodes only have 1 output. However, the mean
can at least be scheduled into the same loop as the variance now since
there is no data dependency. Thus we can take fewer passes over the
data.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy ngimel yf225 chenyang78 kadeng

[ghstack-poisoned]
